File: src/main/java/de/schramm/royalbash/controller/service/core/creature/NoOpCreature.java
package de.schramm.royalbash.controller.service.core.creature;

import de.schramm.royalbash.controller.service.core.CardOnPlayerContext;
import de.schramm.royalbash.controller.service.core.Creature;
import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;

@Value
@Builder(toBuilder = true)
public class NoOpCreature implements Creature {

    private final String id;
    private final String name;
    private final int hitpoints;
    private final int attack;
    private final int cost;
    private final boolean isPlaceableOnSpot = true;

    @Override
    public Game invoke(CardOnPlayerContext cardOnPlayerContext) {
        return CreatureUtil.spawnCreature(this, cardOnPlayerContext);
    }

    @Override
    public Creature damage(Creature attacker) {
        return this.toBuilder()
                .hitpoints(hitpoints - attacker.getAttack())
                .build();
    }

    @Override
    public boolean isDead() {
        return CreatureUtil.isDead(this);
    }
}

File: backend/src/main/kotlin/de/schramm/royalbash/domain/card/creature/NoOpCreature.kt
package de.schramm.royalbash.domain.card.creature

import de.schramm.royalbash.domain.Context
import de.schramm.royalbash.domain.Creature
import de.schramm.royalbash.domain.Game
import de.schramm.royalbash.domain.spawnCreature

data class NoOpCreature(
        override val id: String,
        override val hitpoints: Int,
        override val attack: Int,
        override val cost: Int
) : Creature {

    override val name = "NoOpCreature"
    override val text = "NoOp Creature Text"
    override val image = "NoOp Image URL"

    override fun invoke(context: Context): Game {
        return this.spawnCreature(context)
    }

    override fun damage(amountOfDamage: Int): Creature {
        return this.copy(hitpoints = hitpoints - amountOfDamage)
    }
}
